[ICE]: did not expect duplicate `ConstParamHasTy` for `N/#1` in param-env: ParamEnv {

<!--
[31mICE[0m: Rustc ./a.rs '' 'thread 'rustc' ($TID) panicked at compiler/rustc_middle/src/ty/sty.rs:367:9: 'did not expect duplicate `ConstParamHasTy` for `N/#1` in param-env: ParamEnv {'', 'thread 'rustc' ($TID) panicked at compiler/rustc_middle/src/ty/sty.rs:367:9: 'did not expect duplicate `ConstParamHasTy` for `N/#1` in param-env: ParamEnv {''
File: /tmp/im/redu/a.rs
-->

auto-reduced (treereduce-rust):
````rust
trait Trait<'a, T, const F: bool> {
    fn r#static() -> Self;
}
impl<'T, const N: usize> W<N> {
    reuse Trait::<'T, W<N>, N>::r#static;
}

````

original:
````rust
#![feature(fn_delegation)]
struct W<const N: usize>;
trait Trait<'a, T, const F: bool> {
    fn r#static() -> Self;
}
impl<'T, const N: usize> W<N> {
    reuse Trait::<'T, W<N>, N>::r#static;
}

fn main() {}

````

Version information
````
rustc 1.98.0-nightly (4c9d2bfe4 2026-07-01)
binary: rustc
commit-hash: 4c9d2bfe4ad7a65669098754964aaebe0ec1ced2
commit-date: 2026-07-01
host: x86_64-unknown-linux-gnu
release: 1.98.0-nightly
LLVM version: 22.1.8

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/4c9d2bfe4ad7a65669098754964aaebe0ec1ced2/compiler/rustc_middle/src/ty/sty.rs#L361-L373

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0425]: cannot find type `W` in this scope
 --> /tmp/icemaker_global_tempdir.zm1FldSVy35z/rustc_testrunner_tmpdir_reporting.BimMTowxrpho/mvce.rs:4:26
  |
4 | impl<'T, const N: usize> W<N> {
  |                          ^ not found in this scope

error[E0425]: cannot find type `W` in this scope
 --> /tmp/icemaker_global_tempdir.zm1FldSVy35z/rustc_testrunner_tmpdir_reporting.BimMTowxrpho/mvce.rs:5:23
  |
5 |     reuse Trait::<'T, W<N>, N>::r#static;
  |                       ^ not found in this scope

error[E0658]: functions delegation is not yet fully implemented
 --> /tmp/icemaker_global_tempdir.zm1FldSVy35z/rustc_testrunner_tmpdir_reporting.BimMTowxrpho/mvce.rs:5:5
  |
5 |     reuse Trait::<'T, W<N>, N>::r#static;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #118212 <https://github.com/rust-lang/rust/issues/118212> for more information
  = help: add `#![feature(fn_delegation)]` to the crate attributes to enable
  = note: this compiler was built on 2026-07-01; consider upgrading it if it is out of date

error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.zm1FldSVy35z/rustc_testrunner_tmpdir_reporting.BimMTowxrpho/mvce.rs:6:2
  |
6 | }
  |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.zm1FldSVy35z/rustc_testrunner_tmpdir_reporting.BimMTowxrpho/mvce.rs`


thread 'rustc' (1011618) panicked at /rustc-dev/4c9d2bfe4ad7a65669098754964aaebe0ec1ced2/compiler/rustc_middle/src/ty/sty.rs:367:9:
did not expect duplicate `ConstParamHasTy` for `N/#1` in param-env: ParamEnv {
    caller_bounds: [
        Binder { value: ConstArgHasType(N/#1, bool), bound_vars: [] },
        Binder { value: TraitPredicate(<{type error} as std::marker::Sized>, polarity:Positive), bound_vars: [] },
        Binder { value: TraitPredicate(<{type error} as std::marker::MetaSized>, polarity:Positive), bound_vars: [] },
        Binder { value: ConstArgHasType(N/#1, usize), bound_vars: [] },
    ],
}
stack backtrace:
   0:     0x7f9e1f0e6ac6 - <<std[16495cdcf3c2a980]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[e500fd9f2ab8424]::fmt::Display>::fmt
   1:     0x7f9e1f80f288 - core[e500fd9f2ab8424]::fmt::write
   2:     0x7f9e1f0fbeac - <std[16495cdcf3c2a980]::sys::stdio::unix::Stderr as std[16495cdcf3c2a980]::io::Write>::write_fmt
   3:     0x7f9e1f0baf2a - std[16495cdcf3c2a980]::panicking::default_hook::{closure#0}
   4:     0x7f9e1f0d9213 - std[16495cdcf3c2a980]::panicking::default_hook
   5:     0x7f9e1e0c6108 - std[16495cdcf3c2a980]::panicking::update_hook::<alloc[99712308a9e7a3db]::boxed::Box<rustc_driver_impl[6fdcfad033932331]::install_ice_hook::{closure#1}>>::{closure#0}
   6:     0x7f9e1f0d96b2 - std[16495cdcf3c2a980]::panicking::panic_with_hook
   7:     0x7f9e1f0bafe2 - std[16495cdcf3c2a980]::panicking::panic_handler::{closure#0}
   8:     0x7f9e1f0af4c9 - std[16495cdcf3c2a980]::sys::backtrace::__rust_end_short_backtrace::<std[16495cdcf3c2a980]::panicking::panic_handler::{closure#0}, !>
   9:     0x7f9e1f0bc94d - __rustc[32109bd59933d253]::rust_begin_unwind
  10:     0x7f9e1c07f7bc - core[e500fd9f2ab8424]::panicking::panic_fmt
  11:     0x7f9e211a0ad2 - <rustc_middle[320e6cb0af7a3edc]::ty::sty::ParamConst>::find_const_ty_from_env.cold
  12:     0x7f9e1fec465d - <rustc_trait_selection[82eb004aec24fc7d]::traits::fulfill::FulfillProcessor as rustc_data_structures[850f084c2feb1379]::obligation_forest::ObligationProcessor>::process_obligation
  13:     0x7f9e1f805ad1 - <rustc_data_structures[850f084c2feb1379]::obligation_forest::ObligationForest<rustc_trait_selection[82eb004aec24fc7d]::traits::fulfill::PendingPredicateObligation>>::process_obligations::<rustc_trait_selection[82eb004aec24fc7d]::traits::fulfill::FulfillProcessor>
  14:     0x7f9e1fae6ea6 - <rustc_hir_typeck[e15413bd17dde5e5]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  15:     0x7f9e1fb044eb - <rustc_hir_typeck[e15413bd17dde5e5]::fn_ctxt::FnCtxt>::check_expr_call
  16:     0x7f9e1fae6913 - <rustc_hir_typeck[e15413bd17dde5e5]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  17:     0x7f9e1fabc7e3 - <rustc_hir_typeck[e15413bd17dde5e5]::fn_ctxt::FnCtxt>::check_expr_block
  18:     0x7f9e1fae68ed - <rustc_hir_typeck[e15413bd17dde5e5]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  19:     0x7f9e1faa420f - rustc_hir_typeck[e15413bd17dde5e5]::check::check_fn
  20:     0x7f9e2067d277 - rustc_hir_typeck[e15413bd17dde5e5]::typeck_with_inspect::{closure#0}
  21:     0x7f9e2067bc70 - rustc_query_impl[e8ce24fe214a5db4]::query_impl::typeck_root::invoke_provider_fn::__rust_begin_short_backtrace
  22:     0x7f9e1ff86129 - rustc_query_impl[e8ce24fe214a5db4]::execution::try_execute_query::<rustc_data_structures[850f084c2feb1379]::vec_cache::VecCache<rustc_span[d4ad0aa40b44abc2]::def_id::LocalDefId, rustc_middle[320e6cb0af7a3edc]::query::erase::ErasedData<[u8; 8usize]>, rustc_middle[320e6cb0af7a3edc]::dep_graph::graph::DepNodeIndex>, false>
  23:     0x7f9e1ff85b83 - rustc_query_impl[e8ce24fe214a5db4]::query_impl::typeck_root::execute_query_non_incr::__rust_end_short_backtrace
  24:     0x7f9e1fe35e6d - <rustc_middle[320e6cb0af7a3edc]::ty::context::TyCtxt>::par_hir_body_owners::<rustc_hir_analysis[81978af84cf663ed]::check_crate::{closure#1}>::{closure#0}
  25:     0x7f9e1fe357ab - rustc_hir_analysis[81978af84cf663ed]::check_crate
  26:     0x7f9e1ff86e3a - rustc_interface[8b4b822bfe1350c5]::passes::analysis
  27:     0x7f9e209828c7 - rustc_query_impl[e8ce24fe214a5db4]::execution::try_execute_query::<rustc_middle[320e6cb0af7a3edc]::query::caches::SingleCache<rustc_middle[320e6cb0af7a3edc]::query::erase::ErasedData<[u8; 0usize]>>, false>
  28:     0x7f9e20982529 - rustc_query_impl[e8ce24fe214a5db4]::query_impl::analysis::execute_query_non_incr::__rust_end_short_backtrace
  29:     0x7f9e209b8bc0 - rustc_interface[8b4b822bfe1350c5]::interface::run_compiler::<(), rustc_driver_impl[6fdcfad033932331]::run_compiler::{closure#0}>::{closure#1}
  30:     0x7f9e209915ba - std[16495cdcf3c2a980]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[8b4b822bfe1350c5]::util::run_in_thread_with_globals<rustc_interface[8b4b822bfe1350c5]::util::run_in_thread_pool_with_globals<rustc_interface[8b4b822bfe1350c5]::interface::run_compiler<(), rustc_driver_impl[6fdcfad033932331]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  31:     0x7f9e2099136d - <std[16495cdcf3c2a980]::thread::lifecycle::spawn_unchecked<rustc_interface[8b4b822bfe1350c5]::util::run_in_thread_with_globals<rustc_interface[8b4b822bfe1350c5]::util::run_in_thread_pool_with_globals<rustc_interface[8b4b822bfe1350c5]::interface::run_compiler<(), rustc_driver_impl[6fdcfad033932331]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[e500fd9f2ab8424]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  32:     0x7f9e20995790 - <std[16495cdcf3c2a980]::sys::thread::unix::Thread>::new::thread_start
  33:     0x7f9e1a2981b9 - <unknown>
  34:     0x7f9e1a31d21c - <unknown>
  35:                0x0 - <unknown>

error: the compiler unexpectedly panicked. This is a bug

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.98.0-nightly (4c9d2bfe4 2026-07-01) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [typeck_root] type-checking `<impl at /tmp/icemaker_global_tempdir.zm1FldSVy35z/rustc_testrunner_tmpdir_reporting.BimMTowxrpho/mvce.rs:4:1: 4:30>::r#static`
#1 [analysis] running analysis passes on crate `mvce`
end of query stack
error: aborting due to 4 previous errors

Some errors have detailed explanations: E0425, E0601, E0658.
For more information about an error, try `rustc --explain E0425`.

```

</p>
</details>

<!--
query stack:
        Binder { value: ConstArgHasType(N/#1, bool), bound_vars: [] },
        Binder { value: TraitPredicate(<{type error} as std::marker::Sized>, polarity:Positive), bound_vars: [] },
        Binder { value: TraitPredicate(<{type error} as std::marker::MetaSized>, polarity:Positive), bound_vars: [] },
        Binder { value: ConstArgHasType(N/#1, usize), bound_vars: [] },
#0 [typeck_root] type-checking `<impl at /tmp/icemaker_global_tempdir.zm1FldSVy35z/rustc_testrunner_tmpdir_reporting.BimMTowxrpho/mvce.rs:4:1: 4:30>::r#static`
#1 [analysis] running analysis passes on crate `mvce`

-->
@rustbot label +F-fn_delegation 


<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_hir_analysis/src/delegation.rs
+++ b/compiler/rustc_hir_analysis/src/delegation.rs
@@ -4,7 +4,7 @@
 
 use std::debug_assert_matches;
 
-use rustc_data_structures::fx::FxHashMap;
+use rustc_data_structures::fx::{FxHashMap, FxHashSet};
 use rustc_hir::def::DefKind;
 use rustc_hir::def_id::{DefId, LocalDefId};
 use rustc_hir::{DelegationSelfTyPropagationKind, PathSegment};
@@ -21,6 +21,7 @@ type RemapTable = FxHashMap<u32, u32>;
 struct ParamIndexRemapper<'tcx> {
     tcx: TyCtxt<'tcx>,
     remap_table: RemapTable,
+    delegation_parent_consts: FxHashSet<ty::ParamConst>,
 }
 
 impl<'tcx> TypeFolder<TyCtxt<'tcx>> for ParamIndexRemapper<'tcx> {
@@ -337,7 +338,7 @@ fn create_generic_args<'tcx>(
     delegation_id: LocalDefId,
     mut parent_args: &[ty::GenericArg<'tcx>],
     mut child_args: &[ty::GenericArg<'tcx>],
-) -> Vec<ty::GenericArg<'tcx>> {
+) -> (Vec<ty::GenericArg<'tcx>>, &'tcx [ty::GenericArg<'tcx>]) {
     let delegation_generics = tcx.generics_of(delegation_id);
     let delegation_args = ty::GenericArgs::identity_for_item(tcx, delegation_id);
 
@@ -389,7 +390,7 @@ fn create_generic_args<'tcx>(
     let zero_self = zero_self.as_ref().into_iter();
     let after_lifetimes_self = after_lifetimes_self.as_ref().into_iter();
 
-    zero_self
+    let args = zero_self
         .chain(delegation_parent_args)
         .chain(parent_args.iter().filter(|a| a.as_region().is_some()))
         .chain(child_args.iter().filter(|a| a.as_region().is_some()))
@@ -398,7 +399,9 @@ fn create_generic_args<'tcx>(
         .chain(child_args.iter().filter(|a| a.as_region().is_none()))
         .chain(synth_args)
         .copied()
-        .collect::<Vec<_>>()
+        .collect::<Vec<_>>();
+
+    (args, delegation_parent_args)
 }
 
 pub(crate) fn inherit_predicates_for_delegation_item<'tcx>(
@@ -432,6 +435,44 @@ pub(crate) fn inherit_predicates_for_delegation_item<'tcx>(
                     && trait_pred.self_ty().skip_binder().is_param(0)
                 {
                     continue;
+                }
+
+                // If we have a constant in parent or child args that came from delegation
+                // parent:
+                // ```rust
+                // trait Trait<T, const B: bool> { /* .. */}
+                // impl<const N: usize> S<N> {
+                //     reuse Trait::<S<N>, N>::foo;
+                // }
+                // ```
+                // Then if we inherit const predicate from `Trait` then we end up with
+                // two `ConstArgHasType` for `N` constant:
+                // 1) ConstArgHasType(N/#0, bool) from `Trait`
+                // 2) ConstArgHasType(N/#0, usize) from delegation parent
+                // So in case the constant came from delegation parent we will not inherit
+                // ConstArgHasType from signature.
+                // The check is so complicated because we build generic args for signature
+                // and predicates inheritance, for the example above it will be
+                // `args = [S<N/#0>, N/#0, S<N/#0>, N/#0]`, where
+                // args[0] - Self type, args[1] - delegation parent const, args[2] - first
+                // arg of callee path, args[3] - second arg of callee path.
+                // When processing predicate ConstArgHasType(B/#2, bool)
+                // from delegation signature (`Trait::foo`), we need to map `B/#2` into some
+                // arg from `args`. The mapping which is built by `create_mapping` function is:
+                // `{0: 0, 2: 3, 1: 2}`, so as `B/#2` has index `2` it is mapped into third
+                // arg from `args` - `N/#0`. After we obtained mapped const param, we check if
+                // it came from delegation parent, and if so we do not process its `ConstArgHasType`
+                // predicate.
+                // (Issue #158675).
+                if let ty::PredicateKind::Clause(ty::ClauseKind::ConstArgHasType(ct, _)) =
+                    pred.0.as_predicate().fold_with(&mut self.folder).kind().skip_binder()
+                {
+                    let unnorm_const = EarlyBinder::bind(self.tcx, ct).instantiate(self.tcx, args);
+                    if let ty::ConstKind::Param(param) = unnorm_const.skip_norm_wip().kind()
+                        && self.folder.delegation_parent_consts.contains(&param)
+                    {
+                        continue;
+                    }
                 }
 
                 let new_pred = pred.0.fold_with(&mut self.folder);
@@ -497,10 +538,21 @@ fn create_folder_and_args<'tcx>(
     parent_args: &'tcx [ty::GenericArg<'tcx>],
     child_args: &'tcx [ty::GenericArg<'tcx>],
 ) -> (ParamIndexRemapper<'tcx>, Vec<ty::GenericArg<'tcx>>) {
-    let args = create_generic_args(tcx, sig_id, def_id, parent_args, child_args);
+    let (args, delegation_parent_args) =
+        create_generic_args(tcx, sig_id, def_id, parent_args, child_args);
+
     let remap_table = create_mapping(tcx, sig_id, def_id);
 
-    (ParamIndexRemapper { tcx, remap_table }, args)
+    let delegation_parent_consts = delegation_parent_args
+        .iter()
+        .filter_map(|a| {
+            a.as_const().and_then(|c| {
+                if let ty::ConstKind::Param(param) = c.kind() { Some(param) } else { None }
+            })
+        })
+        .collect();
+
+    (ParamIndexRemapper { tcx, remap_table, delegation_parent_consts }, args)
 }
 
 fn check_constraints<'tcx>(

--- a/tests/ui/delegation/generics/const-predicates-ice-158675.rs
+++ b/tests/ui/delegation/generics/const-predicates-ice-158675.rs
@@ -1,0 +1,30 @@
+#![feature(fn_delegation)]
+
+mod original_ice {
+    struct S<const N: usize>;
+
+    trait Trait<T, const B: bool> {
+        fn fun();
+    }
+
+    impl<const N: usize> S<N> {
+        reuse Trait::<S<N>, N>::fun;
+        //~^ ERROR: the constant `N` is not of type `bool`
+    }
+}
+
+mod with_child_constants {
+    struct S<const N: usize>;
+
+    trait Trait<T, const B: bool> {
+        fn fun<const C: char>();
+    }
+
+    impl<const N: usize> S<N> {
+        reuse Trait::<S<N>, N>::fun::<N>;
+        //~^ ERROR: the constant `N` is not of type `bool`
+        //~| ERROR: the constant `N` is not of type `char`
+    }
+}
+
+fn main() {}

--- a/tests/ui/delegation/generics/const-predicates-ice-158675.stderr
+++ b/tests/ui/delegation/generics/const-predicates-ice-158675.stderr
@@ -1,0 +1,42 @@
+error: the constant `N` is not of type `bool`
+  --> $DIR/const-predicates-ice-158675.rs:11:29
+   |
+LL |         reuse Trait::<S<N>, N>::fun;
+   |                             ^ expected `bool`, found `usize`
+   |
+note: required by a const generic parameter in `original_ice::Trait::fun`
+  --> $DIR/const-predicates-ice-158675.rs:6:20
+   |
+LL |     trait Trait<T, const B: bool> {
+   |                    ^^^^^^^^^^^^^ required by this const generic parameter in `Trait::fun`
+LL |         fn fun();
+   |            --- required by a bound in this associated function
+
+error: the constant `N` is not of type `bool`
+  --> $DIR/const-predicates-ice-158675.rs:24:29
+   |
+LL |         reuse Trait::<S<N>, N>::fun::<N>;
+   |                             ^ expected `bool`, found `usize`
+   |
+note: required by a const generic parameter in `with_child_constants::Trait::fun`
+  --> $DIR/const-predicates-ice-158675.rs:19:20
+   |
+LL |     trait Trait<T, const B: bool> {
+   |                    ^^^^^^^^^^^^^ required by this const generic parameter in `Trait::fun`
+LL |         fn fun<const C: char>();
+   |            --- required by a bound in this associated function
+
+error: the constant `N` is not of type `char`
+  --> $DIR/const-predicates-ice-158675.rs:24:39
+   |
+LL |         reuse Trait::<S<N>, N>::fun::<N>;
+   |                                       ^ expected `char`, found `usize`
+   |
+note: required by a const generic parameter in `with_child_constants::Trait::fun`
+  --> $DIR/const-predicates-ice-158675.rs:20:16
+   |
+LL |         fn fun<const C: char>();
+   |                ^^^^^^^^^^^^^ required by this const generic parameter in `Trait::fun`
+
+error: aborting due to 3 previous errors
+

